x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -0,0 +1,307 @@
+            this.delegatedBlobStore.close();
+        }
+
+        @Override
+        public BlobContainer blobContainer(BlobPath path) {
+            BlobPath encryptionMetadataBlobPath = BlobPath.cleanPath();
+            encryptionMetadataBlobPath = encryptionMetadataBlobPath.add(ENCRYPTION_METADATA_PREFIX + keyId(this.masterSecretKey));
+            for (String pathComponent : path) {
+                encryptionMetadataBlobPath = encryptionMetadataBlobPath.add(pathComponent);
+            }
+            return new EncryptedBlobContainerDecorator(this.delegatedBlobStore.blobContainer(path),
+                    this.delegatedBlobStore.blobContainer(encryptionMetadataBlobPath), this.masterSecretKey);
+        }
+    }
+
+    private static class EncryptedBlobContainerDecorator implements BlobContainer {
+
+        private final BlobContainer delegatedBlobContainer;
+        private final BlobContainer encryptionMetadataBlobContainer;
+        private final SecretKey masterSecretKey;
+
+        EncryptedBlobContainerDecorator(BlobContainer delegatedBlobContainer, BlobContainer encryptionMetadataBlobContainer,
+                SecretKey masterSecretKey) {
+            this.delegatedBlobContainer = delegatedBlobContainer;
+            this.encryptionMetadataBlobContainer = encryptionMetadataBlobContainer;
+            this.masterSecretKey = masterSecretKey;
+        }
+
+        @Override
+        public BlobPath path() {
+            return this.delegatedBlobContainer.path();
+        }
+
+        @Override
+        public InputStream readBlob(String blobName) throws IOException {
+            final BytesReference dataDecryptionKeyBytes = Streams.readFully(this.encryptionMetadataBlobContainer.readBlob(blobName));
+            final SecretKey dataDecryptionKey;
+            try {
+                dataDecryptionKey = unwrapKey(BytesReference.toBytes(dataDecryptionKeyBytes), this.masterSecretKey);
+            } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException e) {
+                throw new IOException(e);
+            }
+            return GCMPacketsCipherInputStream.getDecryptor(this.delegatedBlobContainer.readBlob(blobName), dataDecryptionKey, NONCE,
+                    BC_FIPS_PROVIDER);
+        }
+
+        @Override
+        public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+            final byte[] wrappedDataEncryptionKey;
+            final SecretKey dataEncryptionKey;
+            try {
+                dataEncryptionKey = generateRandomSecretKey();
+                wrappedDataEncryptionKey = wrapKey(dataEncryptionKey, this.masterSecretKey);
+            } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | IllegalBlockSizeException e) {
+                throw new IOException(e);
+            }
+            try (InputStream stream = new ByteArrayInputStream(wrappedDataEncryptionKey)) {
```

Comment:
I think we have to be careful with the ordering here.
We currently delete the metadata first. That means we must write it last. Otherwise, we  lose all consistency guarantees. If we find a metadata blob we should have that imply that a data blob exists IMO. If it doesn't we can't do a safe listing operation on either the metadata nor the data blobs (because a data blob could be the remnant of a half-failed delete and a meta blob could be the remnant of a half-failed write).
If we write meta last and delete it first, then meta blobs can be used as a safe proxy for listing out data blobs and it all works out nicely right? (see my comment on the list methods also)

Reply:
See my [prev comment](https://github.com/elastic/elasticsearch/pull/48221#discussion_r337751497) . It depends which direction for the relation between the two blobs we pick, so that we tolerate half-writes and half-deletes. I think I've been consistent with my choosing but let's discuss it if you disagree.

Reply:
Big sorry here for the noise! I read the delete method the wrong way around (though you deleted the meta before the data) and all the other comments resulted from that mistake.
-> I think your solution is just fine :)

There is one potential upside though that I see to basing it all on the meta (that made me suggest going with the inverse order instead of this one):

Currently adding one meta blob for each blob we write is quite a bit of overhead. A solution that groups the metadata in some way may be a bit tricky as a step 1 (and with the current APIs we have for the blobstore repo) but I think it's not something we should rule out in subsequent steps. Once the work on the metadata handling consistency to fix remaining S3 issues and https://github.com/elastic/elasticsearch/issues/45736 finishes we'll have more options here I think.
If we use the metadata as the source of truth and allow dangling data blobs so long as they're not referenced by other snapshot level metadata I think that leaves us a lot more options in lowering the number of puts and deletes without the tricky (BwC wise) move from data blobs to meta blobs first going forward. Admittedly, that's a bit of a random and non-concrete point and if you think it's easier to go with the current order I'm sure we'll find a BwC solution down the road when we get to optimising things :)